### PR TITLE
[SPARK-40585][SQL] Double-quoted identifiers should be available only in ANSI mode

### PR DESCRIPTION
**Pull request by gengliangwang**
### What changes were proposed in this pull request?

https://github.com/apache/spark/pull/38022 introduces an optional feature for supporting double-quoted identifiers. The feature is controlled by a flag `spark.sql.ansi.double_quoted_identifiers` which is independent from the flag `spark.sql.ansi.enabled`.
This is inconsistent with another ANSI SQL feature "Enforce ANSI reserved keywords": https://spark.apache.org/docs/latest/sql-ref-ansi-compliance.html#sql-keywords-optional-disabled-by-default, which is only available when `spark.sql.ansi.enabled` is true.

Thus, to make the ANSI flags consistent, I suggest making double-quoted identifiers only available under ANSI SQL mode.
Other than that, this PR renames it from `spark.sql.ansi.double_quoted_identifiers` to `spark.sql.ansi.doubleQuotedIdentifiers`
### Why are the changes needed?

To make the ANSI SQL related features consistent.

### Does this PR introduce _any_ user-facing change?

No, the feature is not released yet.

### How was this patch tested?

New SQL test input file under ANSI mode.

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala [hunk 1]
```diff
@@ -2964,14 +2972,6 @@ object SQLConf {
       .booleanConf
       .createWithDefault(false)
 
-  val ENFORCE_RESERVED_KEYWORDS = buildConf("spark.sql.ansi.enforceReservedKeywords")
```
Comment (gengliangwang) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:2967]: Note related to this PR, but I move this code next to the ANSI_ENABLED flag for grouping the ANSI SQL related flags.

**Comment by gengliangwang** [comment]
cc @srielau @entong as well

File: sql/core/src/test/resources/sql-tests/results/double-quoted-identifiers.sql.out [hunk 2]
```diff
@@ -277,74 +277,110 @@ SELECT 1 FROM "not_exist"
 -- !query schema
 struct<>
 -- !query output
-org.apache.spark.sql.AnalysisException
-Table or view not found: not_exist; line 1 pos 14
+org.apache.spark.sql.catalyst.parser.ParseException
```
Comment (cloud-fan) [inline sql/core/src/test/resources/sql-tests/results/double-quoted-identifiers.sql.out:280]: hmm, do we really need to run this SQL gold file in non-ansi mode? We will just see a bunch of parser errors.

Comment (gengliangwang) [reply sql/core/src/test/resources/sql-tests/results/double-quoted-identifiers.sql.out:280]: I intend to ensure it won't work when ANSI mode is off.

Comment (amaliujia) [reply sql/core/src/test/resources/sql-tests/results/double-quoted-identifiers.sql.out:280]: I am seeing there is an extra testing here:

so this test file itself contains both `double_quoted_identifiers=false` and `double_quoted_identifiers=true` and now we run it into both ANSI and non-ANSI.

So I think what the unique testing coverage is:
1. non-ANSI and `double_quoted_identifiers=false`, so double quoted is still a string.
2. non-ANSI and `double_quoted_identifiers=true`, so we see parser exception.
3. ANSI and `double_quoted_identifiers=true`, this feature is on and being tested.

But for ANSI and `double_quoted_identifiers=false` which seems to be the same as number 2 above?

Comment (gengliangwang) [reply sql/core/src/test/resources/sql-tests/results/double-quoted-identifiers.sql.out:280]: Again, the test set is small. I don't see the downside of testing all the combinations.
Let's consider about reducing the tests when it becomes bigger, which is quite unlikely.

Comment (cloud-fan) [reply sql/core/src/test/resources/sql-tests/results/double-quoted-identifiers.sql.out:280]: I think it's time to refactor it now, as we run it again in the ansi mode tests.

It's not just unnecessary test time, but also brings confusion: why do we test `double_quoted_identifiers=false` in ansi mode?

Comment (gengliangwang) [reply sql/core/src/test/resources/sql-tests/results/double-quoted-identifiers.sql.out:280]: > but also brings confusion: why do we test double_quoted_identifiers=false in ansi mode?

Do you mean "in non-ansi" mode?

Comment (gengliangwang) [reply sql/core/src/test/resources/sql-tests/results/double-quoted-identifiers.sql.out:280]: @cloud-fan so we at least need to test the 3 cases below:
* non-ansi mode and double-quoted id enabled
* ansi mode and double-quoted id enabled
* ansi mode and double-quoted id disabled

Any suggestion on the refactor?

Comment (cloud-fan) [reply sql/core/src/test/resources/sql-tests/results/double-quoted-identifiers.sql.out:280]: We can have a `double-quoted-identifiers.sql` without duplicate queries. In `.../inputs/ansi/`, we add 2 files to import the golden file with different configs
```
// file double-quoted-identifiers-enabled.sql
--SET spark.sql.ansi.double_quoted_identifiers = true
--IMPORT double-quoted-identifiers.sql
```
```
// file double-quoted-identifiers-disabled.sql
--SET spark.sql.ansi.double_quoted_identifiers = false
--IMPORT double-quoted-identifiers.sql
```

Comment (cloud-fan) [reply sql/core/src/test/resources/sql-tests/results/double-quoted-identifiers.sql.out:280]: > non-ansi mode and double-quoted id enabled

I don't think so. This is a subconfig if ansi mode, and we only need to test the default case in non-ansi mode.

Comment (gengliangwang) [reply sql/core/src/test/resources/sql-tests/results/double-quoted-identifiers.sql.out]: @cloud-fan Thanks for the suggestion. I updated the tests.

**Review by cloud-fan — APPROVED** [review]

**Comment by gengliangwang** [comment]
Merging to master

**Pull request closed without merging**